pkg/controller/autoops/api_key.go

```diff
@@ -0,0 +1,394 @@
+
+package autoops
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	autoopsv1alpha1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/autoops/v1alpha1"
+	esv1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/elasticsearch/v1"
+	commonesclient "github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/esclient"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/hash"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/metadata"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/reconciler"
+	esclient "github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/client"
+	esuser "github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/user"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/k8s"
+	ulog "github.com/elastic/cloud-on-k8s/v3/pkg/utils/log"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/maps"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/net"
+)
+
+const (
+	// managedByMetadataKey is the metadata key indicating the API key is managed by ECK.
+	// This is used when storing the API key in Elasticsearch to clearly identify it as managed by ECK.
+	// This is not included in the labels of the secret.
+	managedByMetadataKey = "elasticsearch.k8s.elastic.co/managed-by"
+	// managedByValue is the value for the managed-by metadata
+	managedByValue = "eck"
+	// configHashMetadataKey is the metadata key storing the hash API key
+	configHashMetadataKey = "elasticsearch.k8s.elastic.co/config-hash"
+	// PolicyNameLabelKey is the label key for the AutoOpsAgentPolicy name.
+	// This is exported as its used in the remotecluster controller to identify API keys managed by the autoops controller.
+	PolicyNameLabelKey = "autoops.k8s.elastic.co/policy-name"
+	// policyNamespaceLabelKey is the label key for the AutoOpsAgentPolicy namespace
+	policyNamespaceLabelKey = "autoops.k8s.elastic.co/policy-namespace"
+)
+
+// apiKeySpec represents the specification for an autoops API key
+type apiKeySpec struct {
+	roleDescriptors map[string]esclient.Role
+}
+
+// reconcileAutoOpsESAPIKey reconciles the API key and secret for a specific Elasticsearch cluster.
+func reconcileAutoOpsESAPIKey(
+	ctx context.Context,
+	c k8s.Client,
+	esClientProvider commonesclient.Provider,
+	dialer net.Dialer,
+	policy autoopsv1alpha1.AutoOpsAgentPolicy,
+	es esv1.Elasticsearch,
+) error {
+	log := ulog.FromContext(ctx).WithValues(
+		"es_namespace", es.Namespace,
+		"es_name", es.Name,
+	)
+	log.V(1).Info("Reconciling AutoOps ES API key")
+
+	esClient, err := esClientProvider(ctx, c, dialer, es)
+	if err != nil {
+		return fmt.Errorf("while creating Elasticsearch client for %s/%s: %w", es.Namespace, es.Name, err)
+	}
+	defer esClient.Close()
+
+	stackMonitoringUserRole, ok := esuser.PredefinedRoles[esuser.StackMonitoringUserRole].(esclient.Role)
```

Comment:
This gives more permissions than necessary I believe. We should follow what is documented here https://www.elastic.co/docs/deploy-manage/monitor/autoops/cc-connect-self-managed-to-autoops